walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,154 @@
+  * a collection of TimeRanges in which the MeetingRequest can be satisfied given the
+  * the constraints posed by the attendees' events. 
+  */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ArrayList<String> attendeesWithoutEvents = new ArrayList<>(request.getAttendees());
+    int initialAttendeeCount = attendeesWithoutEvents.size();
+    ArrayList<String> optionalAttendeesWithoutEvents = new ArrayList<>(request.getOptionalAttendees());
+    int initialOptionalAttendeeCount = optionalAttendeesWithoutEvents.size();
+    
+    // Stores all TimeRanges of mandatory attendees with no repetitions.
+    Collection<TimeRange> timesRequiredAttendees = new HashSet<>();
+    // Stores all TimeRanges of optional attendees with no repetitions.
+    Collection<TimeRange> timesOptionalAttendees = new HashSet<>();
+    // Stores all TimeRanges in which mandatory attendees are busy with no overlaps.
+    ArrayList<TimeRange> notAvailableTimes = new ArrayList<>();
+    // Stores all TimeRanges in which optional attendees are busy with no overlaps.
+    ArrayList<TimeRange> notOptionalTimes = new ArrayList<>();
+    // Stores all TimeRanges in which mandatory attendees can meet.
+    ArrayList<TimeRange> availableTimes = new ArrayList<>();
+    // Stores all TimeRanges in which optional attendees can meet.
+    ArrayList<TimeRange> optionalCalendar = new ArrayList<>();
+
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+    // No events to conflict with.
+    if (events.size() == 0) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+      
+    ArrayList<String> toDeleteMan = new ArrayList<>();
+    ArrayList<String> toDeleteOpt = new ArrayList<>();
+      
+    for (Event event : events) {
+      for (String attendee : event.getAttendees()) {
+        if (attendeesWithoutEvents.contains(attendee)) {
+          // Adds TimeRange if any mandatory attendee attends this event.
+          timesRequiredAttendees.add(event.getWhen());
+          // Adds whenever a mandatory attendee appears in an event. 
+          toDeleteMan.add(attendee);
+        }
+        if (optionalAttendeesWithoutEvents.contains(attendee)) {
+          timesOptionalAttendees.add(event.getWhen());
+          toDeleteOpt.add(attendee);
+        }
+      }
+    }
+    // Deletes from our initial attendee counts those who appears on events so we can check 
+    // whether there is no overlap between the attendees of our given Events and those request in Meeting Request.
+    attendeesWithoutEvents.removeAll(toDeleteMan);
+    optionalAttendeesWithoutEvents.removeAll(toDeleteOpt);
+    
+    // Transforming collections of TimeRanges into an ArrayList to allow indexing.
+    ArrayList<TimeRange> allTimesRequiredAttendees = new ArrayList<>(timesRequiredAttendees);
```

Comment:
Ahh the comment clears up what's going on here. I would note though that with clearer variable names you probably wouldn't need a comment to distinguish what the difference is between "allTimesRequiredAttendees" and "timesRequiredAttendees". (Neither name gives a clue that one is a list and the other is a collection.) Not something that necessarily requires fixing but definitely something to keep in mind for the capstone project.